### PR TITLE
fix get retention step for reverse series

### PR DESCRIPTION
**Pull request by msaf1980**
Fix for https://github.com/lomik/graphite-clickhouse/issues/142

### CODE REVIEW

File: pkg/alias/map.go [hunk 1]
```diff
@@ -70,16 +70,18 @@ func (m *Map) Size() int {
 }
 
 // Series returns keys of aliases map
-func (m *Map) Series(isReverse bool) []string {
+func (m *Map) Series(isReverse bool) ([]string, []string) {
```
Comment (Felixoid) [inline pkg/alias/map.go:73]: It doubles memory consumption for metric names of each request. Maybe we'd try to fix the current broken place?

Comment (msaf1980) [reply pkg/alias/map.go:73]: Yes, double memory (for metric names only). I can try some refactor and avoid this.

File: render/query.go [hunk 2]
```diff
@@ -128,7 +128,7 @@ func FetchDataPoints(ctx context.Context, cfg *config.Config, fetchRequests Mult
 		if tf.MaxDataPoints <= 0 {
 			tf.MaxDataPoints = int64(cfg.ClickHouse.MaxDataPoints)
 		}
-		targets.pointsTable, targets.isReverse, targets.rollupObj = SelectDataTable(cfg, tf.From, tf.Until, targets.List, chContext)
+		targets.pointsTable, targets.isReverse, targets.rollupUseReverted, targets.rollupObj = SelectDataTable(cfg, tf.From, tf.Until, targets.List, chContext)
```
Comment (Felixoid) [inline render/query.go:131]: I already feel like it makes sense to return `*config.DataTable, error` from SelectDataTable. And then change `render.Targets` accordingly. Does it look reasonable?

File: render/query.go [hunk 3]
```diff
@@ -218,11 +218,24 @@ func (r *Reply) getDataAggregated(ctx context.Context, cfg *config.Config, tf Ti
 	data = EmptyData
 
 	metricList := targets.AM.Series(targets.isReverse)
```
Comment (Felixoid) [inline render/query.go:220]: This is used in `queryCarbonlink(ctx, cfg, metricList)` and should be always in direct order.

```suggestion
	metricList := targets.AM.Series(false)
```

See following suggestion

File: render/query.go [hunk 4]
```diff
@@ -218,11 +218,24 @@ func (r *Reply) getDataAggregated(ctx context.Context, cfg *config.Config, tf Ti
...
+		metricListUnreverse = metricList
+	}
+
+	var metricListRuleLookup []string
+	if targets.isReverse && targets.rollupUseReverted {
+		metricListRuleLookup = metricListUnreverse
+	} else {
+		metricListRuleLookup = metricList
+	}
+
```
Comment (Felixoid) [inline render/query.go:238]: We need only one additional slice
```suggestion
	var metricListRuleLookup []string
	if targets.isReverse {
		metricListRuleLookup = targets.AM.Series(true)
	} else {
		metricListRuleLookup = metricList
	}

```

File: render/query.go [hunk 5]
```diff
@@ -249,8 +262,9 @@ func (r *Reply) getDataAggregated(ctx context.Context, cfg *config.Config, tf Ti
 		}
 
 		if targets.isReverse {
```
Comment (Felixoid) [inline render/query.go:264]: No need `if` here, just `m = metricList[n]` when suggestions are accepted

**Review by Felixoid — COMMENTED** [review]
Some thoughts regarding new code

File: render/query.go (hunk 2 above)
Comment (msaf1980) [reply render/query.go:131]: Yes, like targets.isReverse. It's per table config option. So, returned from SelectDataTable.

File: render/query.go (hunk 3 above)
Comment (msaf1980) [reply render/query.go:220]: Why ? Now it's used for Path list in clickhouse query. So it must to coexistense with table type.
At master for carbonlink used metricList (which can be reverse). So, set to

`carbonlinkResponseRead := queryCarbonlink(ctx, cfg, metricListUnreverse)`

File: render/query.go (hunk 4 above)
Comment (msaf1980) [reply render/query.go:238]: > We need only one additional slice

It's not so simple.  

We need reverse or unreverse options  on some cases, like
`for n, m := range metricListRuleLookup {` - m is used for roollup lookup

We need unreverse  on some cases, like (for store in steps)
```
if targets.isReverse {
	m = metricListUnreverse[n]
}
```

File: render/query.go (hunk 5 above)
Comment (msaf1980) [reply render/query.go:264]: After this step, m must be unreversed for use in steps map. See code after 
https://github.com/lomik/graphite-clickhouse/blob/master/render/query.go#L375

File: render/query.go (hunk 2 above)
Comment (msaf1980) [reply render/query.go:131]: Also reorg SelectDataTable (as method of Targets struct), return error only. All required fileds set in method directly.

File: render/query.go [hunk 6]
```diff
@@ -344,7 +344,7 @@ func (r *Reply) getDataAggregated(ctx context.Context, cfg *config.Config, tf Ti
 	}
 
 	carbonlinkData := carbonlinkResponseRead()
-	data, err = parseAggregatedResponse(ctx, b, e, carbonlinkData, targets.isReverse)
+	data, err = parseAggregatedResponse(ctx, b, e, carbonlinkData, false)
```
Comment (Felixoid) [inline render/query.go:347]: This should stay as it was, `targets.isReverse` is used to reverse metric names only from ClickHouse

Comment (msaf1980) [reply render/query.go:347]: I's carbonlink response parse. Not Clickhouse.

Comment (Felixoid) [reply render/query.go:347]: Here and in `getDataUnaggregated` `parse*ggregateResponse` reads everything. `carbonlinkData` is `extraPoints`, so parsed as is. `targets.isReverse` is used in parsing of ClickHouse response in `Scan()`

File: render/query.go [hunk 7]
```diff
@@ -239,7 +251,7 @@ func (r *Reply) getDataAggregated(ctx context.Context, cfg *config.Config, tf Ti
 
 	// Grouping metrics by aggregation steps and functions
 	var step int64
-	for _, m := range metricList {
+	for n, m := range metricListRuleLookup {
 		newStep, agg := targets.rollupObj.Lookup(m, uint32(age))
 		step = r.cStep.calculateUnsafe(step, int64(newStep))
 		if mm, ok := bodyAggregation[agg.Name()]; ok {
```
Comment (Felixoid) [inline render/query.go:257]: 258 + 260 are broken now. There must be metricList elements.

Comment (msaf1980) [reply render/query.go:257]: Why ? It's a start point of this PR.  I pass reverse or unreverse metric name to rule lookup based on table config settings.
```
if targets.isReverse && targets.rollupUseReverted {
	metricListRuleLookup = metricListUnreverse
} else {
	metricListRuleLookup = metricList
}
```

Comment (Felixoid) [reply render/query.go:257]: I talk about `m` in lines 258 + 260 specifically. Instead, there should be `metricList[n]`

Comment (Felixoid) [reply render/query.go:257]: As a reference, see https://github.com/lomik/graphite-clickhouse/pull/143/files#diff-dd0047c86eb3449b47b84a87cbc1cf2bbbb726459e13916af2a1586abcea4399R414, it stays unchanged, and bodyAggregation should use metricList as it was before

**Comment by Felixoid** [comment]
Thank you for your contribution!

**Pull request merged**